src/federation.cpp

```diff
@@ -227,6 +237,173 @@ static bool ReadFederationOpts() {
+//! A collection of valid signatures for the proposed block of the current round.
+static std::map<CPubKey, std::vector<unsigned char>> g_block_sigs;
+
+static bool GenerateBlockProposal()
+{
+    const CChainParams& params = Params();
+    const Consensus::Params& consensusParams = params.GetConsensus();
+
+    // FIXME: Is this mandatory coinbase destination really necessary?  I recall
+    //        that this is from the Liquid codebase to ensure that Blockstream
+    //        collected fees on Liquid rather than the functionaries.  Is this
+    //        something we want to carry over to our chain?  If so, there are
+    //        probably better ways to accomplish fee capture.
+    CScript fee_destination_script = consensusParams.mandatory_coinbase_destination;
+    // If no forced fee destination script is specified in the chain parameters,
+    // then it is not consensus-controlled.  For testing purposes, we make it
+    // anyone-can-spend.  Let's change this later when we decide what to do
+    // about collecting fees.
+    if (fee_destination_script == CScript()) {
+        fee_destination_script = CScript() << OP_TRUE;
+    }
+
+    // The required_wait parameter is useful for preventing DoS of the
+    // federation itself.  By only using transactions which have been in the
+    // mempool for X seconds, we ensure that there is probably sufficient time
+    // for those transactions to have reached the other federation nodes before
+    // the start of the round.
+    std::chrono::seconds required_wait{5}; // FIXME: fetch this from config?
+
+    std::unique_ptr<CBlockTemplate> pblocktemplate(BlockAssembler(*g_context->mempool, params).CreateNewBlock(fee_destination_script, required_wait));
+    if (!pblocktemplate.get()) {
+        LogPrint(BCLog::FEDERATION, "Error generating block template proposal.\n");
+        return false;
+    }
+
+    CBlock &block = pblocktemplate->block;
+    {
+        // IncrementExtraNonce sets coinbase flags and builds the merkle tree
+        LOCK(cs_main);
+        unsigned int extra_nonce = 0;
+        IncrementExtraNonce(&block, g_context->chainman->ActiveTip(), extra_nonce);
+    }
+
+    LOCK(cs_block_signer);
+
+    using std::swap;
+    swap(g_block_proposal, block);
+    return true;
+}
+
+// Verifies that the proposed block is building off the tip and is a valid
+// block, then signs a relayable confirmation that we will sign this block and
+// accept it if the signing threshold is reached.
+static bool AcceptBlockProposal()
+{
+    const CChainParams& params = Params();
+
+    LOCK(cs_block_signer);
+    const CBlock& block = g_block_proposal;
+
+    LOCK(cs_main);
+
+    // Verify that this isn't a replay of a prior block.
+    BlockMap::iterator mi = g_context->chainman->BlockIndex().find(block.GetHash());
+    if (mi != g_context->chainman->BlockIndex().end()) {
+        LogPrint(BCLog::FEDERATION, "Already have proposed block: %s\n", block.GetHash().GetHex());
+        return false;
```

Comment:
I don't think this particular `return false` will have any effect. Right now the only time the return value is checked is when the leader of the round checks their own block proposal. And barring hash collisions that should never trigger this code path. We don't use it for punishing peers.

The check is more along the lines of "is this proposal one we're willing to sign?" And a `return false` just terminates participation in the round.

Reply:
I modified the other call to `AcceptBlockProposal` so it checks the return value too. Since returning here (true or false) means we aren't signing the block proposal and the round terminates (for this peer), I think `return false` is correct. It saves a few cycles in that case.